### DISCUSSION TITLE
GKE K8S autoscaling

### DISCUSSION THREAD

**Discussion by baszalmstra** [cloudnative-pg/cloudnative-pg#1552] [General]
Im running Cloudnative-PG on Google managed kubernetes engine. Very happy with this project so far!

One thing I notice though is that GKE is unable to scale _down_ nodes that have an active database cluster pod running on them.

The reason for not being able to move the pod is `no.scale.down.node.pod.not.backed.by.controller`. GKE also complains about a notification: "Pod is blocking scale down because it’s not backed by a controller".

Is there a way to fix this?

**Comment by pratheekrebala** [comment]
We have been running into this as well. We are also on GKE but I believe this affects all K8s clusters using auto-scaling.  This is caused by the Pod Disruption Budget that is created for the pods running the primary instance. This PDB targets pods with the label `role: primary`. This means, only primary pods will block a scale-down.

FWIW, I think the current design is a good default that prevents a lot of footguns associated with auto-scaling and with spot-instances.

**Some Solutions**

- One trivial fix here is to add the following annotation to the `podTemplate` which allows auto-scaler to ignore the usual protection: `cluster-autoscaler.kubernetes.io/safe-to-evict: true`
  - You can do this through the `inheritedMetadata` property on the `ClusterSpec`:

      ```
      inheritedMetadata:
        annotations:
          cluster-autoscaler.kubernetes.io/safe-to-evict: true
      ```

- Another option is to update the PDB to allow for an unhealthy primary.
- **Ideal**: Place your primaries on a non auto-scaling node pool and put the replicas on an auto-scaling pool. 


This is discussed in more detail in the [Operator Capability Levels](https://cloudnative-pg.io/documentation/1.18/operator_capability_levels/) section of the operator documentation.

Additional information about the auto-scaling behavior can be found in the [K8s auto-scaler documentation.](
https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#what-types-of-pods-can-prevent-ca-from-removing-a-node)

**Comment by pratheekrebala** [comment]
I forgot to mention, this is happening because the auto-scaler does not recognize that the pods are managed by a custom controller and assumes they are not replicated. The operator does add an `ownerReference` indicating that the pod is managed by an external controller.

However, the auto-scaler has a [hard-coded list of controllers](https://github.com/kubernetes/autoscaler/blob/ee667550c081d99924e7b1100f0c1596551be3d6/cluster-autoscaler/utils/drain/drain.go#L124-L204): `ReplicationController`, `Daemonsets`, `Jobs`, `ReplicaSet` and `Statefulset`.

There's an open issue to allow the auto-scaler to acknowledge custom controllers: https://github.com/kubernetes/autoscaler/issues/5387 and a potential PR https://github.com/kubernetes/autoscaler/pull/5419

**Comment by phisco** [comment]
Hi, thank you very much @pratheekrebala for your thorough answer! 
To add to what you said, it's true that the PDB is blocking the primary instance from being evicted, but if the primary instance is running on an unschedulable node the operator triggers a switchover to an instance located on a schedulable node, actually unlocking the nkde the primary was previously running on. This allows rollout upgrades 1 node at a time to work properly and maybe could also cover this use-case. 
I mean that if setting the annotation you mentioned above means the pod is not seen as an issue and a normal drain is started, then it shouldn't create any major disruption except a switchover and that the pod's eviction might take a little bit more time than expected.

**Comment by sxd** [comment]
this should be a discussion, looks like we can create a bug from this but not yet

**Reply by aran** [reply]
Given the upstream k8s update detailed below, does it make sense to open a refined issue based on this?

**Comment by baszalmstra** [comment]
Thanks for the detailed answer! Im trying to understand this properly.

Adding the following metadata to the cluster spec.

```
inheritedMetadata:
  annotations:
    cluster-autoscaler.kubernetes.io/safe-to-evict: true
```

will allow cluster autoscaler to evict pods. Which would resolve the `no.scale.down.node.pod.not.backed.by.controller` issue?

If I have 1 primary and 1 replica running is that safe? Or could this result in downtime?

**Reply by shauryagoel** [reply]
Hi, were you able to try it? I am getting `no.scale.down.node.pod.not.enough.pdb` issue after using the above annotation.

**Comment by BadLiveware** [comment]
Facing the same issue on GKE, correct me if I am wrong but the `cluster-autoscaler.kubernetes.io/safe-to-evict: true` doesnt bypass PDBs.

I dont really understand why the PDB is selecting only the primary node. Running with 2+ replicas I would very much like a lever to pull to allow pods to be moved. My specific use case is for small dbs(~20GB) that get up and running quickly, I can see this being undesirable for larger dbs.

**Comment by dananichev** [comment]
Anyways, autoscaler repo indicates https://github.com/kubernetes/autoscaler/pull/5507 needed feature was added into it. 

Any progress on working solution for such use cases? It is quite tiresome to perform small k8s clusters upgrades.

**Comment by aran** [comment]
Might be a way to integrate the target label of the PDB with something derived from the cluster replication settings when synchronous replication is enabled.
